File: 2020/05-binary-boarding/main.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strings"
	"time"
)

func decode(seat string) (row, col int) {
	in := [2]int{0, 127}

	for _, c := range seat[:7] {
		if c == 'F' {
			in[1] -= (in[1] - in[0]) / 2
			in[1]--
		} else {
			in[0] += (in[1] - in[0]) / 2
			in[0]++
		}
	}
	row = in[0]

	in = [2]int{0, 7}
	for _, c := range seat[7:] {
		if c == 'L' {
			in[1] -= (in[1] - in[0]) / 2
			in[1]--
		} else {
			in[0] += (in[1] - in[0]) / 2
			in[0]++
		}
	}
	col = in[0]

	return
}

func solve(f string) (p1, p2 int) {
	in, _ := os.ReadFile(f)
	lines := strings.Split(strings.TrimSpace(string(in)), "\n")

	seats := []int{}
	for _, line := range lines {
		row, col := decode(line)
		seat := row*8 + col
		seats = append(seats, seat)
	}

	for i := slices.Min(seats); i <= slices.Max(seats); i++ {
		if !slices.Contains(seats, i) {
			return slices.Max(seats), i
		}
	}

	return
}

func main() {
	start := time.Now()
	fmt.Println("Day 05: Binary Boarding")
	p1, p2 := solve("input.txt")
	fmt.Println("\tPart One:", p1) // 978
	fmt.Println("\tPart Two:", p2) // 727
	fmt.Printf("Done in %.3f seconds \n", time.Since(start).Seconds())
}

File: 2020/05-binary-boarding/main.py
import time, os, math

# binary partitioning of interval
def decode(interval, line, chr):
	for c in line:
		if c == chr:
			interval[1] -= math.floor((interval[1] - interval[0])/2)
			interval[1] -=1
		else:
			interval[0] += math.floor((interval[1] - interval[0])/2)
			interval[0] +=1
	return interval[0]

def solve(file="input.txt"):
	p1,p2 = 0,0
	lines = open(file, "r").read().strip().split("\n")
	lines = [line.strip() for line in lines]

	seats = set()
	for line in lines:
		row = decode([0,127], line[:7], 'F')
		col = decode([0,7], line[7:], 'L')
		seats.add(row*8 + col)

	# Part 2
	for seat in range(min(seats), max(seats)):
		if seat not in seats:
			return max(seats),seat

os.chdir(__file__.rsplit("/", 1)[0])
print("Day 05: Binary Boarding")
start = time.time()
p1, p2 = solve("input.txt")
print("\tPart 1: ", p1)
print("\tPart 2: ", p2)
# assert p1 == 978
# assert p2 == 727
print("Done in %s seconds" % round(time.time() - start, 3))
